contracts/PartyBuy.sol

```diff
@@ -0,0 +1,170 @@
+
+    // the timestamp at which the Party is no longer active
+    uint256 public expiresAt;
+    // the maximum price that the party is willing to
+    // spend on the token
+    // NOTE: the party can accept *UP TO* 102.5% of maxPrice in total,
+    // and will not accept more contributions after this
+    uint256 public maxPrice;
+
+    // ============ Events ============
+
+    // emitted when the token is successfully bought
+    event Bought(address triggeredBy, address targetAddress, uint256 ethSpent, uint256 ethFeePaid, uint256 totalContributed);
+
+    // emitted if the Party fails to buy the token before expiresAt
+    // and someone expires the Party so folks can reclaim ETH
+    event Expired(address triggeredBy);
+
+    // ======== Constructor =========
+
+    constructor(
+        address _partyDAOMultisig,
+        address _tokenVaultFactory,
+        address _weth,
+        address _allowList
+    ) Party(_partyDAOMultisig, _tokenVaultFactory, _weth) {
+        allowList = IAllowList(_allowList);
+    }
+
+    // ======== Initializer =========
+
+    function initialize(
+        address _nftContract,
+        uint256 _tokenId,
+        uint256 _maxPrice,
+        uint256 _secondsToTimeout,
+        Structs.AddressAndAmount calldata _split,
+        Structs.AddressAndAmount calldata _tokenGate,
+        string memory _name,
+        string memory _symbol
+    ) external initializer {
+        // validate maxPrice
+        require(_maxPrice > 0, "PartyBuy::initialize: must set price higher than 0");
+        // initialize & validate shared Party variables
+        __Party_init(_nftContract, _tokenId, _split, _tokenGate, _name, _symbol);
+        // set PartyBuy-specific state variables
+        expiresAt = block.timestamp + _secondsToTimeout;
+        maxPrice = _maxPrice;
+    }
+
+    // ======== External: Contribute =========
+
+    /**
+     * @notice Contribute to the Party's treasury
+     * while the Party is still active
+     * @dev Emits a Contributed event upon success; callable by anyone
+     */
+    function contribute() external payable nonReentrant {
+        // require that the new total contributed is not greater than
+        // the maximum amount the Party is willing to spend
+        require(totalContributedToParty + msg.value <= getMaximumContributions(), "PartyBuy::contribute: cannot contribute more than max");
+        // continue with shared _contribute flow
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::buy: party not active"
+        );
```

Comment:
Don't want to distract from other comments / more expert opinions here but: I think from our side, it's okay if someone triggers a buy after the expiration block stamp. Some parties might want to try to keep going after that time. In which case, "expiration" might be a confusing name, but idk. Just adding in case that's helpful.

Reply:
yep! what John said. it's ideal if the Party is able to buy the token. expiring the Party is meant to give an escape hatch if that's impossible, but it doesn't have to be a "hard stop"